[JSC] Use parseRecursively for eval preprocess

Constellation opened the pull request:
#### 65ae8f7a2d0ac2f644fb495b441364f09ea822f5
<pre>
[JSC] Use parseRecursively for eval preprocess
<a href="https://bugs.webkit.org/show_bug.cgi?id=288947">https://bugs.webkit.org/show_bug.cgi?id=288947</a>
<a href="https://rdar.apple.com/145983863">rdar://145983863</a>

Reviewed by Keith Miller.

This patch accepts parseRecursively for SloppyJSON parsing mode too.

* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::requires):
(JSC::reviverMode&gt;::parse):
(JSC::reviverMode&gt;::tryJSONPParse): Deleted.
(JSC::reviverMode&gt;::parseRecursivelyEntry): Deleted.
(JSC::reviverMode&gt;::parseRecursively): Deleted.
* Source/JavaScriptCore/runtime/LiteralParser.h:
(JSC::LiteralParser::requires):
(JSC::LiteralParser::tryLiteralParse): Deleted.
(JSC::LiteralParser::tryLiteralParsePrimitiveValue): Deleted.

Canonical link: <a href="https://commits.webkit.org/291509@main">https://commits.webkit.org/291509@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fbd01626d394e4bcde2aef3e60edaa732430abf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93117 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2375 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98116 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43643 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12950 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21125 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71193 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28608 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96119 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/9752 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84251 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51521 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9445 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1885 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42956 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/85827 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79736 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1869 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100143 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91783 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20169 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80218 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20421 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80154 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79522 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24064 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13264 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14891 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20153 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25329 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114432 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19840 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33016 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23300 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/21581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -64,6 +64,8 @@ inline const CharType* LiteralParser<CharType, reviverMode>::Lexer::currentToken
 bool LiteralParser<CharType, reviverMode>::tryJSONPParse(Vector<JSONPData>& results, bool needsFullSourceInfo)
 {
+    ASSERT(m_mode == JSONP);
+    ASSERT(reviverMode == JSONReviverMode::Disabled);
```
Since `JSONReviverMode` is a constant template argument, let's make this a `static_assert` instead.

Constellation replied:
This does not work since reviverMode `Enabled` case is also instantiated.

MenloDorian commented on Source/JavaScriptCore/runtime/LiteralParser.h:
```diff
@@ -151,25 +151,34 @@ class LiteralParser {
     JSValue tryLiteralParse()
     {
+        ASSERT(m_mode == StrictJSON);
+        ASSERT(reviverMode == JSONReviverMode::Disabled);
```
Ditto.  Make this a `static_assert` on `reviverMode`.

Constellation replied:
Ditto.

MenloDorian commented on Source/JavaScriptCore/runtime/LiteralParser.h:
```diff
@@ -151,25 +151,34 @@ class LiteralParser {
     JSValue tryLiteralParse(JSONRanges* sourceRanges)
     {
+        ASSERT(m_mode == StrictJSON);
+        ASSERT(reviverMode == JSONReviverMode::Enabled);
```
Ditto.  Make this a `static_assert` on `reviverMode`.

Constellation replied:
Ditto

MenloDorian commented on Source/JavaScriptCore/runtime/LiteralParser.h:
```diff
@@ -151,25 +151,34 @@ class LiteralParser {
+    JSValue tryEval()
+    {
+        ASSERT(m_mode == SloppyJSON);
+        ASSERT(reviverMode == JSONReviverMode::Disabled);
```
Ditto.  Make this a `static_assert` on `reviverMode`.

Constellation replied:
Ditto.

MenloDorian commented on Source/JavaScriptCore/runtime/LiteralParser.h:
```diff
@@ -180,6 +189,7 @@ class LiteralParser {
     JSValue tryLiteralParsePrimitiveValue()
     {
         ASSERT(m_mode == StrictJSON);
+        ASSERT(reviverMode == JSONReviverMode::Disabled);
```
Ditto.  Make this a `static_assert` on `reviverMode`.

Constellation replied:
Ditto.

MenloDorian commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -1263,15 +1265,28 @@ JSValue LiteralParser<CharType, reviverMode>::parseRecursivelyEntry(VM& vm)
 template<typename CharType, JSONReviverMode reviverMode>
+JSValue LiteralParser<CharType, reviverMode>::evalRecursivelyEntry(VM& vm)
+{
+    ASSERT(m_mode == SloppyJSON);
```
Since this is only called from `tryEval`, please also add a `static_assert(reviverMode == JSONReviverMode::Disabled));` here to document this invariant.

Constellation replied:
Ditto.

MenloDorian commented on Source/JavaScriptCore/runtime/LiteralParser.h:
```diff
@@ -151,25 +151,34 @@ class LiteralParser {
+        ASSERT(reviverMode == JSONReviverMode::Enabled);
         m_lexer.next();
-        JSValue result = parse(getVM(m_globalObject), m_mode == StrictJSON ? StartParseExpression : StartParseStatement, sourceRanges);
+        JSValue result = parse(getVM(m_globalObject), StartParseExpression, sourceRanges);
```
Can you explain the relationship between `m_mode == StrictJSON` and `StartParseExpression : StartParseStatement`?  Why does `StrictJSON` imply `Statement`?  What does `ParseStatement` mean in this context?  Thanks.

Constellation replied:
`StartParseStatement` means it can accept `(` and `)`, which is not usable for JSON.

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -1257,21 +1259,37 @@ ALWAYS_INLINE JSValue LiteralParser<CharType, reviverMode>::parsePrimitiveValue(
 
 template<typename CharType, JSONReviverMode reviverMode>
 JSValue LiteralParser<CharType, reviverMode>::parseRecursivelyEntry(VM& vm)
+    requires (reviverMode == JSONReviverMode::Disabled)
```
Nit: Technically you don't need these at the definition, I think?

kmiller68 reviewed (approved):
r=me with nits.

Constellation committed:
[JSC] Use parseRecursively for eval preprocess
https://bugs.webkit.org/show_bug.cgi?id=288947
rdar://145983863

Reviewed by Keith Miller.

This patch accepts parseRecursively for SloppyJSON parsing mode too.

* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::requires):
(JSC::reviverMode>::parse):
(JSC::reviverMode>::tryJSONPParse): Deleted.
(JSC::reviverMode>::parseRecursivelyEntry): Deleted.
(JSC::reviverMode>::parseRecursively): Deleted.
* Source/JavaScriptCore/runtime/LiteralParser.h:
(JSC::LiteralParser::requires):
(JSC::LiteralParser::tryLiteralParse): Deleted.
(JSC::LiteralParser::tryLiteralParsePrimitiveValue): Deleted.

Canonical link: https://commits.webkit.org/291509@main
